django/db/backends/postgresql/base.py

```diff
@@ -179,6 +197,45 @@ class DatabaseWrapper(BaseDatabaseWrapper):
     ops_class = DatabaseOperations
     # PostgreSQL backend-specific attributes.
     _named_cursor_idx = 0
+    _connection_pools = {}
+
+    @property
+    def pool(self):
+        if not is_psycopg3:
+            return None
+
+        if self.alias not in self._connection_pools:
```

Comment:
In my code, I used `(alias, dbname, conn_params)` as the key, not just the alias. I don't remember why, but I don't think it was to fix something in particular.

One thing that using only `alias` does mean is that any changes made to the `DATABASES` entry after the fact will be completely ignored, although this is probably not something that Django condones officially (maybe in tests?).

Reply:
Mhm, yes we should probably connect this to `setting_changed` signal and then clean up `self._connection_pools`. But aside from that I think the alias should be enough, changing settings is really really really not supported ;)

Reply:
One production scenario where this is relevant:

I (used to) use ["AWS RDS IAM authentication"](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.Python.html) (similar in [Google](https://cloud.google.com/sql/docs/postgres/iam-logins#logging-in-as-a-user), [Microsoft](https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/how-to-connect-with-managed-identity#retrieve-the-access-token-from-the-azure-instance-metadata-service), [Vault](https://developer.hashicorp.com/vault/docs/secrets/databases/postgresql)), basically it means you call an external service and get a temporary connection password that is valid for 15 minutes. The way people usually implement this it is by overriding `get_connection_params` and setting `params['password']`, taking care to refresh the password when needed.

I sent a question about this to psycopg https://github.com/psycopg/psycopg/discussions/500 but since I stopped using this method I haven't followed up unfortunately. However the comment by Daniele does suggest that changing the connection params on the pool after the fact can be officially supported if the work is done to ensure it doesn't regress.

Reply:
That is fair, but how gets `self._connection_pools` ever cleaned up if the `conn_params` change every X minutes? Maybe we should have it as `{alias: (dbname, conn_params, pool)}` and verify that dbname/conn_params didn't change?

Reply:
What are you expecting if the dbname/conn_params have changed? Should I update them in the dictionary, close the existing pool and open a new one?

Reply:
As @apollo13 said, including the params in the key is definitely no good, it will cause `cls._connection_pools` to grow unbounded in the "new password every X minutes" scenario.

As I mentioned in the psycopg discussion link, I think the only viable solution is to update the pool itself on `settings_changed` such that it will use the new params when creating new connections. Currently it's actually possible to do this, but Daniele says it's accidental, and before it can be relied upon, needs a regression test in psycopg.

I think that for now, we should say that changing settings dynamically is not going to work with `pool`.

> What are you expecting if the dbname/conn_params have changed? Should I update them in the dictionary, close the existing pool and open a new one?

I can't see this working; existing connections using the old password/setting continue to work, closing the pool will close all existing connections (even ones currently in use by ongoing requests), which is very disruptive.